WIP: View Spec implementation

c-thiel opened the pull request:
Implementation of the Iceberg V1 ViewMetadata

ZENOTME commented on crates/iceberg/src/catalog/mod.rs:
```diff
@@ -438,6 +438,31 @@ impl TableUpdate {
     }
 }
 
+/// TableCreation represents the creation of a table in the catalog.
```
TableCreation->ViewCreation

c-thiel replied:
Fixed in the new commit together with clips

ZENOTME commented:
Thanks! Nice work! @c-thiel

c-thiel commented:
@ZENOTME one feature I did not implement yet is respecting "version.history.num-entries" as mentioned in the [View Spec](https://iceberg.apache.org/view-spec/#view-metadata). I noticed that the table implementation also doesn't implement it. I don't think it matters much unless one is implementing a catalog, but I wanted to mention it.

Fokko commented:
Thanks for working on this @c-thiel. Out of curiosity, and probably this is a broader discussion, but currently, the view spec is  around [SQL representation](https://iceberg.apache.org/view-spec/#representations). This might not be the most natural fit with dataframe-based systems. Do you have any thoughts on how to represent views in iceberg-rust?

c-thiel commented:
@Fokko thats is a hard topic.

The idealist in me would like to eventually see something like substrait beeing used.
Adoption of the project is very slow across engines though.

Maybe the more practical short-term solution would be to stick to representations and even limit to one dialect - the one of the "writer", including its technology. Then each client can decide if there is a transpiler good enough for that dialect. 
Funnily enough exactly this ambiguity - that there could be different dialects in the representation (say spark and trino) - gave me a lot of headaches. Which should I use if I am Datafusion? From which should I transpile? What to do if they are not identical?

Sticking to multiple representations is in my opinion only an option if we let the catalog handle this complexity.
Unless substrait gets a significant boost, I would assume that the following approach presents the most stable API (Assuming there is only REST Catalog left ;) ):

- A client would create a view as it does today - however limited to a single representation
- Upon read, a client would additionally present a set of "understood" dialects. The catalog can either serve them or it can't.

This way a Rest-Catalog could even switch to Substrait internally in the future and start now with storing the presented SQL "as-is" without beeing able to serve any other dialect.

With SQLMesh coming up and also Substrait continuing there is at least some development in the area.

c-thiel commented:
@Fokko from my side this is good to merge - types are complete and tests are passing.

c-thiel commented:
@Fokko @ZENOTME 
are there any points open from your side that prevent us to merge the View Spec?
If so, please let me know :)

Fokko commented:
> The idealist in me would like to eventually see something like substrait beeing used.

There is an open discussion on evolving the spec into something like that. I think transpiling SQL will never work in the long run because there are so many variations and extensions to the SQL dialect (custom UDFs and such), maybe you can reach 90% of the SQL which is good enough. 

@nastra would you have time to go over this PR?

nastra commented on crates/iceberg/testdata/view_metadata/ViewMetadataV2Valid.json:
```diff
@@ -0,0 +1,58 @@
+{
```
it would be good to have more JSON files for testing (not just the happy case)

nastra replied:
also in general I think this needs more extensive testing. For example on the java side we have https://github.com/apache/iceberg/tree/5ea78e3fbe5d8c9c846a1b86bcd2b77d13a31acd/core/src/test/java/org/apache/iceberg/view.

Specifically in `TestViewMetadata` we have a bunch of test cases that would be good to cover here as well

c-thiel replied:
Ok, thanks for the feedback @nastra. I'll add some more tests next week.

c-thiel replied:
@nastra I added more tests that cover non-happy cases. 
However, there isn't much in https://github.com/apache/iceberg/tree/5ea78e3fbe5d8c9c846a1b86bcd2b77d13a31acd/core/src/test/java/org/apache/iceberg/view we can use for now. The `ViewMetadataBuilder` I started with is on the same level as the `TableMetadataBuilder` - it can do almost nothing except changing the UUID. The functions tested in the java part are simply not implemented yet. We do have a more complete implementation over at our Iceberg catalog and want to PR them here (https://github.com/hansetag/iceberg-catalog/issues/46), I just want to start with the foundation before heading into the more controversial builder bits.
Would this be alright for you?

Besides the Metadata-Builder, I would like to create 2 follow-up issues:
- Currently in `Snapshots` we use a fallible method for the timestamp() method. For views I am using a non-fallible version, creating a small inconsistency. I would create a new issues to remove `unwrap` from timestamps where they are used.
- Allow for empty namespaces, following our discussion below. I am not sure exactly how to handle this, but right now Rust does not allow the creation of the empty Namespace as used in spark (https://github.com/apache/iceberg-rust/blob/48f9e3e8bff0224569da8dd5beb89ed0c8cc4513/crates/iceberg/src/catalog/mod.rs#L111).

nastra replied:
> However, there isn't much in https://github.com/apache/iceberg/tree/5ea78e3fbe5d8c9c846a1b86bcd2b77d13a31acd/core/src/test/java/org/apache/iceberg/view we can use for now. 

In the java code we typically don't construct the full JSON just to test something, but there are plenty of tests in `TestViewMetadata` to make sure schema / view version re-assignment / de-duplication happens or SQL dialect deduplication behaves correctly. 

> I just want to start with the foundation before heading into the more controversial builder bits.

I was merely asked to look at this view implementation since I did the Java implementation, but I'm not involved in iceberg-rust, so I'll let others decide how to proceed with adding views to iceberg-rust and what to include in this PR vs in follow-ups. 
Regarding follow-ups, I think a few more things would be needed for the view metadata builder, which I mentioned above. `TestViewMetadata´ gives a good overview over how the view metadata / builder should behave.

Thanks for working on this btw 💯

c-thiel replied:
Ok, thanks Eduard for the Feedback.

My preferred way of going forward would be:

1. Merge this PR which just contains the structs & (de)serialization (very much like the `TableMetadaData` in its current state)
2. Figure out a good way for the `TableMetadataBuilder` including partition binding. We have a first shot ready that we can create a PR for.
3. Use the same pattern we use for the `TableMetadataBuilder` for the much lighter `ViewMetadataBuilder` which then includes all the tests Eduard mentioned.

I am aware that a lot of things are missing in the builder. It was a deliberate decision from me to get views up-to-speed with tables first, and then in a second step extend both tables and views features to the java level.

@ZENOTME , @nastra would that be OK for you?

c-thiel commented:
@nastra, @Fokko during testing we found a Problem with the "default-namespace". I am hoping for some insights from your side:
According to the iceberg view spec, "default-namespace" is required: https://iceberg.apache.org/view-spec/#versions.
As such it is modeled as `NamespaceIdent`:
https://github.com/c-thiel/iceberg-rust/blob/d8a7aabce73888e5b22712d60ca95b856db66fc2/crates/iceberg/src/spec/view_version.rs#L56

Valid Namespace Identifiers have at least one element, which is validated by the `NamespaceIdent` constructor.

When creating views, spark creates Metadata objects that specify the "default-namespace" field but as an empty vec. This is a quite unlucky situation, I am not sure what the desired behavior is:
1. My feeling is that this is a bug in the Iceberg Spec. [] as "default-namespace" provides the same information as a non-required field (null), as a namespace without an element is unusable. However, views in spark work, which shows that default-namespace is not always required. Thus it should be optional.
2. It's a bug in the spark implementation and in fact "default-namespace" should always contain at least one element.
3. They are both right, in which case we would except [] as a valid namespace and must change the rust implementation.
4. [] is in general not a valid namespace unless it is used as a "default-namespace" in which some special behavior is expected. This however brings me back to 1 - why not just use null in this case?

nastra commented:
> When creating views, spark creates Metadata objects that specify the "default-namespace" field but as an empty vec

Is this with plain OSS Spark or are you using the iceberg-spark-runtime.jar from Iceberg?

twuebi commented:
Hi @nastra,

this script:

```python3
import pyspark
from pyspark.conf import SparkConf
from pyspark.sql import SparkSession
import pandas as pd

CATALOG_URL = "http://server:8080/catalog"
MANAGEMENT_URL = "http://server:8080/management"
DEMO_WAREHOUSE = "demo"
SPARK_VERSION = pyspark.__version__
SPARK_MINOR_VERSION = '.'.join(SPARK_VERSION.split('.')[:2])
ICEBERG_VERSION = "1.5.2"

config = {
    "spark.sql.catalog.demo-catalog": "org.apache.iceberg.spark.SparkCatalog",
    "spark.sql.catalog.demo-catalog.type": "rest",
    "spark.sql.catalog.demo-catalog.uri": CATALOG_URL,
    "spark.sql.catalog.demo-catalog.warehouse": DEMO_WAREHOUSE,
    "spark.sql.catalog.demo-catalog.io-impl": "org.apache.iceberg.aws.s3.S3FileIO",
    "spark.sql.extensions": "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions",
    "spark.sql.defaultCatalog": "demo-catalog",
    "spark.jars.packages": "org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.5.2,org.apache.iceberg:iceberg-aws-bundle:1.5.2",
}

spark_config = SparkConf().setMaster('local').setAppName("Iceberg-REST")
for k, v in config.items():
    spark_config = spark_config.set(k, v)

spark = SparkSession.builder.config(conf=spark_config).getOrCreate()
spark.sql("CREATE NAMESPACE IF NOT EXISTS spark_demo")
data = pd.DataFrame([[1, 'a-string', 2.2]], columns=['id', 'strings', 'floats'])
sdf = spark.createDataFrame(data)
sdf.writeTo("spark_demo.my_table").createOrReplace()
spark.sql("CREATE view spark_demo.vv as select * from spark_demo.my_table")
```

ends up sending that POST to `/catalog/v1/df910b7e-3912-11ef-a048-6b2680efd54d/namespaces/spark_demo/views`

```json
{
  "name": "vv",
  "schema": {
    "schema-id": 0,
    "type": "struct",
    "fields": [
      {
        "id": 0,
        "name": "id",
        "required": false,
        "type": "long"
      },
      {
        "id": 1,
        "name": "strings",
        "required": false,
        "type": "string"
      },
      {
        "id": 2,
        "name": "floats",
        "required": false,
        "type": "double"
      }
    ]
  },
  "view-version": {
    "version-id": 1,
    "schema-id": 0,
    "timestamp-ms": 1719994069884,
    "summary": {
      "app-id": "local-1719994018459",
      "engine-name": "spark",
      "iceberg-version": "Apache Iceberg 1.5.2 (commit cbb853073e681b4075d7c8707610dceecbee3a82)",
      "engine-version": "3.5.1"
    },
    "representations": [
      {
        "type": "sql",
        "sql": "select * from spark_demo.my_table",
        "dialect": "spark"
      }
    ],
    "default-namespace": []
  },
  "properties": {
    "create_engine_version": "Spark 3.5.1",
    "spark.query-column-names": "id,strings,floats",
    "engine_version": "Spark 3.5.1"
  }
}
```

twuebi commented:
Setting the namespace via:

```python3
spark.sql("USE spark_demo")
spark.sql("CREATE view vv3 as select * from spark_demo.mytable")
```

ends up setting `default-namespace`:


```json
{
  "name": "vv3",
  "schema": {
    "schema-id": 0,
    "type": "struct",
    "fields": [
      {
        "id": 0,
        "name": "id",
        "required": false,
        "type": "long"
      },
      {
        "id": 1,
        "name": "strings",
        "required": false,
        "type": "string"
      },
      {
        "id": 2,
        "name": "floats",
        "required": false,
        "type": "double"
      }
    ]
  },
  "view-version": {
    "version-id": 1,
    "schema-id": 0,
    "timestamp-ms": 1719994988301,
    "summary": {
      "engine-version": "3.5.1",
      "app-id": "local-1719994018459",
      "engine-name": "spark",
      "iceberg-version": "Apache Iceberg 1.5.2 (commit cbb853073e681b4075d7c8707610dceecbee3a82)"
    },
    "representations": [
      {
        "type": "sql",
        "sql": "select * from spark_demo.my_table",
        "dialect": "spark"
      }
    ],
    "default-namespace": [
      "spark_demo"
    ]
  },
  "properties": {
    "spark.query-column-names": "id,strings,floats",
    "engine_version": "Spark 3.5.1",
    "create_engine_version": "Spark 3.5.1"
  }
}
```

nastra commented:
`"default-namespace": []` indicates an empty namespace, not a null one, which is a valid case and there are tests for that here: https://github.com/apache/iceberg/blob/42a2c19cec31c626cbff6cc2dfafb86cdf223bd0/core/src/test/java/org/apache/iceberg/view/TestViewMetadataParser.java#L56. I've also opened https://github.com/apache/iceberg/pull/9890 a while ago to properly test empty namespaces with catalogs that support it.

However, why you're getting an empty namespace when creating the view with a qualified namespace is a different thing (which I need to look into and reproduce on my end)

nastra commented:
Ok I checked the surrounding code and the handling is correct. In https://github.com/apache/iceberg/blob/2a39af894f4f00aa37922ef765cc2583517fa1d1/spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateV2ViewExec.scala#L51 we're always determining the current namespace (which can be empty), which is then what `ViewVersion#defaultNamespace` is set to. As I mentioned earlier that there's a difference between a null namespace and an empty one.

Spark later uses this info in https://github.com/apache/iceberg/blob/6bbf70a52ebccfaba4e7e08facd72b84b571e2a6/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkView.java#L75 in case there's no namespace in the underlying SQL and in your case you fully-qualified `select * from spark_demo.my_table` with a namespace. 

In your second example you configured the current namespace via `USE <namespace>`, which was then evaluated https://github.com/apache/iceberg/blob/2a39af894f4f00aa37922ef765cc2583517fa1d1/spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateV2ViewExec.scala#L51 and passed to `ViewVersion#defaultNamespace` (but not used when actually querying the view).

twuebi commented:
Thanks for checking this @nastra, I understand that we should be relaxing the constraints of NamespaceIdent then if an empty default-namespace is expected.

nastra commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,682 @@
+    pub(crate) versions: HashMap<i64, ViewVersionRef>,
+    /// A list of version log entries with the timestamp and version-id for every
+    /// change to current-version-id
+    pub(crate) version_log: Vec<ViewVersionLog>,
```
in java we call this `history`

c-thiel replied:
Hm, in the iceberg spec its called versions & version-log https://iceberg.apache.org/view-spec/#view-metadata.
I implemented the same way `TableMetadata` is currently implemented: Calling the struct field identically to the spec, but then creating a history accessor.

TableMetadata:
https://github.com/c-thiel/iceberg-rust/blob/ca9de89ac9d95683c8fe9191f72ab922dc4c7672/crates/iceberg/src/spec/table_metadata.rs#L208-L211

ViewMetadata:
https://github.com/c-thiel/iceberg-rust/blob/44630160be1bcf48249c31006b76a7150a029619/crates/iceberg/src/spec/view_metadata.rs#L151-L156

As the `versions` field is not public, it actually implements almost the same interface as java.

@ZENOTME or some other rust dev, it would be great to get some opinion on this.

nastra commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,682 @@
+
+    /// Append view version to view
+    pub fn append_version(&mut self, view_version: ViewVersion) {
+        self.current_version_id = view_version.version_id();
```
I feel that things like view version id re-assignment and view version de-duplication are missing in the builder.
The same applies for schema id re-assignment / schema de-duplication.

There are a bunch of tests in the java ref impl in `TestViewMetadata` that make sure these things behave correctly

c-thiel replied:
Yes, you are right. Bad idea to implement this half-done.
I am removing this function and handle it properly later in the full implementation of the builders.

nastra commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,682 @@
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+#[serde(rename_all = "kebab-case")]
+/// A log of when each snapshot was made.
+pub struct ViewVersionLog {
```
in java this is called `ViewHistoryEntry`, maybe worth naming it here similarly?

c-thiel replied:
I am also happy with renaming.
I think we should use the same naming we settle on in our discussion above. https://github.com/apache/iceberg-rust/pull/331#discussion_r1675578086

nastra commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,347 @@
+}
+
+/// A list of view representations.
+pub type ViewRepresentations = Vec<ViewRepresentation>;
```
is it normal in rust to create a separate type for a list of X?

c-thiel replied:
Normally not. This is looking forward to the ViewMetadataBuilder where we want to implement methods on those.
The motivation is that multiple `ViewRepresentations` are more than just a Vec<ViewRepresentation> - there are many rules in constructing it (i.e. no duplicate dialects). We shouldn't just expose the Vec interface externally because then users could just add an item to the list without going through any of our validation functions (which are not there yet, will come with the ViewMetadataBuilder).

Its a very valuable comment though, because I am still exposing the interface currently. I am putting it in a struct with a private field.

c-thiel replied:
Fix: https://github.com/c-thiel/iceberg-rust/commit/2b427e511dbeb3dfa6292af6983ee8d0b8bc48ad

nastra commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,347 @@
+
+    /// Add a representation to the list.
+    ///
+    /// SQL representations dialects must be unique. If a representation with the same
```
does this make sure to ignore the case of the dialect when comparing two dialects? If no, then it should

c-thiel replied:
No, it did not. I fixed it here:
https://github.com/c-thiel/iceberg-rust/commit/bf95eb7c5c458e01352afd9e5d6199d5c470477e

The behavior is now to ignore the case when adding, but storing the proper case internally (see the test in my commit).
Is this the desired behavior or should we just lowercase it?

nastra commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,347 @@
+    /// Add a SQL representation to the list.
+    ///
+    /// SQL representations dialects must be unique. If a representation with the same
+    /// dialect already exists, it will be overwritten.
```
why is it better to overwrite a dialect rather than throw an error saying that there are duplicate dialects?

c-thiel replied:
I guess it depends on the usage. This method is part of a builder which is not only used to create an object initially, but also to modify an object.
If we fail, there is no way to ever modify a dialect in the builder.
I am adding fallible methods additionally in https://github.com/c-thiel/iceberg-rust/commit/83e4a6f516cc064e7949d2b48ccd615a3b3d0c3c (tests in next commit)

c-thiel commented:
@nastra I hope I addressed all your issues regarding functionality. If you are happy with the implementations, feel free to close our discussions :).
@ZENOTME it would now be great to get some Feedback on the rust side of things. Please check my discussions with Eduard above. There are a few open points regarding naming and also the introduction of types. It would be great to get a review from you or some other Rust Maintainer for this.

c-thiel commented:
@nastra just following up on my last comment - could you close our discussion where you agree with my fix?
@liurenjie1024 could you maybe have a look at the rust side of things? Please check my discussions with Eduard above. There are a few open points regarding naming and also the introduction of types. It would be great to get a review from you or some other Rust Maintainer for this.

c-thiel commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -332,7 +332,7 @@ impl TableMetadataBuilder {
         let table_metadata = TableMetadata {
             format_version: FormatVersion::V2,
-            table_uuid: Uuid::new_v4(),
+            table_uuid: Uuid::now_v7(),
```
This was a long shot. Let me try to explain my motivation. That being said, I could also be convinced to roll back.
I see the following points on each side. Those are small differences, I am though very much on the v7 side.

PRO v7:
- Temporal locality - this is super helpful when looking at S3 and having the default "names" automatically sorted by creation date
- with more REST-Catalogs coming up, tables and views are going to be stored in Databases or KV-Stores with their IDs indexed. It is significantly better to index v7 due to temporal locality than v4 and it reduces the global index size (i.e. https://itnext.io/why-uuid7-is-better-than-uuid4-as-clustered-index-edb02bf70056)
- Conveying information for debug purposes by including the creation timestamp

PRO v4:
- Lets stick to tradition
- The creation timestamp is confidential information that we don't want to leak. If so, v7 is out of the game.

c-thiel commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,675 @@
+    }
+
+    /// Returns the last updated timestamp as a DateTime<Utc> with millisecond precision.
+    pub fn timestamp(self) -> MappedLocalTime<DateTime<Utc>> {
```
Conversion to `DateTime<Utc>` is fallible - thus it would become a `Result`. We already have `timestamp_ms()` as a non-fallible version. Do you still think we should change?

c-thiel replied:
What I absolutely don't like is the `TableMetadata` way of doing this which just panics:
https://github.com/apache/iceberg-rust/blob/f5b406f08f85d8bca402ce840962deee9d6b34c0/crates/iceberg/src/spec/table_metadata.rs#L148

c-thiel replied:
https://github.com/c-thiel/iceberg-rust/commit/d354e63e912cf059f7234c1a0a77f3c3dfb7d219

c-thiel replied:
And my proposal for `TableMetadata`.
https://github.com/c-thiel/iceberg-rust/commit/1c49ff91e68f53c769e1d3df2c1aff8ab6d348a5

Do you have a better Idea for method naming? I think `last_updated_ms` was a really bad method for returning a `DateTime` in the first place. Nonetheless, it now changes the behavior to return an int instead so that we still have a non-fallible accessor.

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,477 @@
+/// A view versions represents the definition of a view at a specific point in time.
+pub struct ViewVersion {
+    /// A unique long ID
+    version_id: i64,
```
Addressed in https://github.com/c-thiel/iceberg-rust/commit/20d9bbb603db1b919fd31247d3b86b86d1911a5d. Is this what you thought of?

We can now store more sort-orders for TableMetadata than View versions. I think the max-values should eventually become part of the iceberg spec.

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,477 @@
+}
+
+/// A builder for [`ViewRepresentations`].
+pub struct ViewRepresentationsBuilder(Vec<ViewRepresentation>);
```
Don't we still need it to create an action?

Should we keep the [`ViewMetadataBuilder`](https://github.com/c-thiel/iceberg-rust/blob/86f1fba39098c55e3295d569964e41005b90a2f6/crates/iceberg/src/spec/view_metadata.rs#L151)?

c-thiel replied:
Ok, I removed the builder in https://github.com/c-thiel/iceberg-rust/commit/06d09ab674f46465293be5dcb58a5af705863bd2.
Is there an issue already for the transaction we could mention this commit in? I assume most of the tests and maybe a bit of the logic might be re-usable.

c-thiel commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,675 @@
+            .with_version_id(initial_version_id)
+            .build();
+
+        let versions = HashMap::from_iter(vec![(initial_version_id, version.into())]);
```
I don't quite understand your point here. Can a schema-id be invalid? Or do you mean rule-out a `-1`?
If so, we should probably make the same changes here right?
https://github.com/c-thiel/iceberg-rust/blob/abd03a7fd66edf6fb83f4f9a98e0e22b4ddac370/crates/iceberg/src/spec/table_metadata.rs#L345

Can you think of more checks?

c-thiel commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,675 @@
+    }
+
+    impl Serialize for ViewMetadata {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
```
Which concrete type do you think we should use? Don't we want to keep this open to allow different serializers? (i.e. json, yaml, maybe a DB Serializer?)

If we change it here, I think we should also change TableMetadata Serialization:
https://github.com/apache/iceberg-rust/blob/f5b406f08f85d8bca402ce840962deee9d6b34c0/crates/iceberg/src/spec/table_metadata.rs#L479-L490

c-thiel commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,675 @@
+        let metadata_builder = ViewMetadataBuilder::new(metadata);
+        let uuid = Uuid::new_v4();
+        let metadata = metadata_builder.assign_uuid(uuid).unwrap().build().unwrap();
+        assert_eq!(metadata.uuid(), uuid);
```
Added it as you suggested in https://github.com/c-thiel/iceberg-rust/commit/b4123fba668b89b149c789dc0f9aff468d7c73f4.
Do you think anything else is missing test-wise?

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,477 @@
+        self.0.into_iter()
+    }
+}
+impl<'a> IntoIterator for &'a ViewRepresentations {
```
Oh, I screwed up the impl, we cannot implement `Iterator` on an immutable object - no place to store the state. However, I just implemented iter() on the object itself.

c-thiel replied:
New version: https://github.com/c-thiel/iceberg-rust/blob/06d09ab674f46465293be5dcb58a5af705863bd2/crates/iceberg/src/spec/view_version.rs#L150-L152

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,477 @@
+    /// This is a helper module that defines types to help with serialization/deserialization.
+    /// For deserialization the input first gets read into either the [SnapshotV1] or [SnapshotV2] struct
+    /// and then converted into the [Snapshot] struct. Serialization works the other way around.
+    /// [SnapshotV1] and [SnapshotV2] are internal struct that are only used for serialization and deserialization.
```
You are right, strange - pressed the GitHub button.
https://github.com/c-thiel/iceberg-rust/commit/ec4facb02b0fea29d36dd88640fd2af8f708bedb

c-thiel commented on crates/iceberg/src/spec/view_metadata.rs:
```diff
@@ -0,0 +1,675 @@
+            default_namespace,
+            summary,
+        } = view_creation;
+        let initial_version_id = super::INITIAL_SEQUENCE_NUMBER;
```
You are right. In Java it seems to be one as well.
Why though? `SortOrder`, `Schema`, `PartitionSpec` all start from 0. Why not views?

c-thiel replied:
Lets stick to 1 then :)

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -155,6 +155,11 @@ impl ViewRepresentations {
     }
+
+    /// Get an iterator over the representations
+    pub fn iter(&self) -> std::slice::Iter<'_, ViewRepresentation> {
```
Oh yes, much better: https://github.com/c-thiel/iceberg-rust/commit/c20c22d8e23667c2fd7a98dce7ee3c1d0da6c9b0

c-thiel commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -71,8 +74,18 @@ impl ViewVersion {
-    pub fn timestamp(&self) -> MappedLocalTime<DateTime<Utc>> {
-        Utc.timestamp_millis_opt(self.timestamp_ms)
+    pub fn timestamp(&self) -> Result<DateTime<Utc>> {
+        match Utc.timestamp_millis_opt(self.timestamp_ms) {
```
Hm, I added it to the error module but I am not happy with it.
https://github.com/c-thiel/iceberg-rust/commit/82dc949d838ea5e3ad769d3cabd3ee4a2262118b

The problem is that the result of `Utc.timestamp_millis_opt(timestamp_ms)` offers no built-in method to convert it to an `Result`. So we cannot use `?` but require a uitility function instead. The one I've written is only remotely related to errors - which is why I don't really like it to be in `errors.rs`. I am not sure though if it's worth adding a new utility module just for timestamp conversions.

Any thoughts?

c-thiel commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -917,8 +933,24 @@ pub struct SnapshotLog {
+            )),
+            MappedLocalTime::None => Err(Error::new(
+                ErrorKind::Unexpected,
+                "Invalid timestamp in snapshot log",
```
https://github.com/c-thiel/iceberg-rust/commit/ea7f2ea9e35be6d4f42e89a9e22938661272e209

c-thiel commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -917,8 +933,24 @@ pub struct SnapshotLog {
+            MappedLocalTime::Single(t) => Ok(t),
+            MappedLocalTime::Ambiguous(_, _) => Err(Error::new(
+                ErrorKind::Unexpected,
+                "Ambiguous timestamp in snapshot log",
```
https://github.com/c-thiel/iceberg-rust/commit/ea7f2ea9e35be6d4f42e89a9e22938661272e209

c-thiel commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -144,8 +144,24 @@ impl TableMetadata {
+            )),
+            MappedLocalTime::None => Err(Error::new(
+                ErrorKind::Unexpected,
+                "Invalid timestamp in table metadata last updated",
```
https://github.com/c-thiel/iceberg-rust/commit/ea7f2ea9e35be6d4f42e89a9e22938661272e209

c-thiel commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -144,8 +144,24 @@ impl TableMetadata {
+            MappedLocalTime::Ambiguous(_, _) => Err(Error::new(
+                ErrorKind::Unexpected,
+                "Ambiguous timestamp in table metadata last updated",
+            )),
```
https://github.com/c-thiel/iceberg-rust/commit/c10c90bc5ddba0233ee1bde150c8c6147b4d72be

Xuanwo commented:
> cc @Xuanwo @nastra @Fokko to take another look

Hi, thanks for the invitation, but I'm not familiar with this topic. Also, apologies for not being involved from the start. I will delegate my decision-making to @Fokko instead.

Xuanwo commented on crates/iceberg/src/error.rs:
```diff
@@ -333,6 +335,28 @@ define_from_err!(
+pub(crate) fn timestamp_ms_to_utc(timestamp_ms: i64) -> Result<DateTime<Utc>> {
+    match Utc.timestamp_millis_opt(timestamp_ms) {
+        chrono::LocalResult::Single(t) => Ok(t),
+        chrono::LocalResult::Ambiguous(_, _) => Err(Error::new(
```
How about print the detailed info in error?

c-thiel replied:
I am not quite sure what do you mean with detailed info in error - we already have the context "timestamp value". Would you add a  more verbose message?

Xuanwo replied:
> I am not quite sure what do you mean with detailed info in error - we already have the context "timestamp value". Would you add a more verbose message?

I want to carry the info returned in `LocalResult::Ambiguous(_, _)`. But I'm guessing they are not useful?

Xuanwo commented on crates/iceberg/src/error.rs:
```diff
@@ -333,6 +335,28 @@ define_from_err!(
+    match Utc.timestamp_millis_opt(timestamp_ms) {
+        chrono::LocalResult::Single(t) => Ok(t),
+        chrono::LocalResult::Ambiguous(_, _) => Err(Error::new(
+            ErrorKind::Unexpected,
```
This this error be `InvalidData`? It's not quiet `Unexpected` to me.

c-thiel replied:
I think what @liurenjie1024 proposed above is most correct:
Ambiguous is unexpected - it would mean there is a fold in UTC time, which sounds unexpected to me.
None on the other hand is returned for different things. Among them is a gap in UTC time, which would be unexpected again, missing timezone data - also unexpected, but most likely an overflow - definitely invalid data.
So I am changing None -> DataInvalid
https://github.com/c-thiel/iceberg-rust/commit/abcb4cd0b664333e594706768ba63ed19ca1d7f6

Xuanwo commented on crates/iceberg/src/spec/mod.rs:
```diff
@@ -38,3 +40,6 @@ pub use sort::*;
 pub use values::*;
+pub use view_metadata::*;
+pub use view_version::*;
+// pub use view_metadata::*;
```
We need to remove it or just uncomment it?

c-thiel replied:
https://github.com/c-thiel/iceberg-rust/commit/ae40cd515254872f1c2da4b0855373e9a7076540

Xuanwo commented on crates/iceberg/src/spec/view_version.rs:
```diff
@@ -0,0 +1,313 @@
+pub enum ViewRepresentation {
+    #[serde(rename = "sql")]
+    /// The SQL representation stores the view definition as a SQL SELECT,
+    SqlViewRepresentation(SqlViewRepresentation),
```
How about use `Sql` as variant name? So users will use it as `ViewRepresentation::Sql` instead of `ViewRepresentation::SqlViewRepresentation`.

c-thiel replied:
Definitely better for usability:
https://github.com/c-thiel/iceberg-rust/commit/bb9cf9c9b290de972fef6953066759a860790cfb

Xuanwo reviewed (commented):
Thank you for your work! I've highlighted a few minor issues from the perspective of rust. Please feel safe to ignore them. I think they are not blockers of merging this PR.

c-thiel commented:
Rebased on current master
